include/seqan3/search/configuration/mode.hpp

```diff
@@ -75,28 +78,166 @@ struct strata : detail::strong_type<uint8_t, strata, detail::strong_type_skill::
+                      "Expected this configuration to be dynamic, i.e. "
+                      "std::same_as<search_mode_t, seqan3::dynamic_state> evaluates to true.");
+        search_mode_variant = std::move(new_search_mode);
+        return *this;
+    }
+    //!\}
+
     //!\privatesection
     //!\brief Internal id to check for consistent configuration settings.
     static constexpr detail::search_config_id id{detail::search_config_id::mode};
+
+    //!\brief Returns the variant containing the search modes.
+    constexpr search_modes_t const & search_modes() const noexcept
+    //!\cond
+        requires std::same_as<search_mode_t, dynamic_state>
+    //!\endcond
+    {
+        return search_mode_variant;
+    }
+
+    //!\brief Evaluates to `true` if the search mode can be configured at runtime, otherwise `false`.
+    static constexpr bool has_dynamic_state = std::same_as<search_mode_t, dynamic_state>;
+
+private:
+    //!\brief Stores the dynamically selected search mode option.
+    search_modes_t search_mode_variant{};
 };
 
 /*!\name Type deduction guides
  * \relates seqan3::search_cfg::mode
  * \{
  */
+//!\brief Deduces the dynamic search mode type.
+mode() -> mode<dynamic_state>;
 
-//!\brief Deduces search mode type from constructor argument.
-template <typename mode_t>
-mode(mode_t) -> mode<remove_cvref_t<mode_t>>;
+//!\brief Deduces search mode option type from constructor argument.
+template <typename search_mode_t>
+mode(search_mode_t) -> mode<search_mode_t>;
 //!\}
 } // namespace seqan3::search_cfg
+
+namespace seqan3::detail
+{
+/*!\brief An internal representation of the associated seqan3::search_cfg::mode.
+ * \ingroup search_configuration
+ *
+ * \tparam search_mode_t The type of the search mode option; must be one of seqan3::detail::search_mode_all,
+ *                       seqan3::detail::search_mode_best, seqan3::detail::search_mode_all_best, or
+ *                       seqan3::detail::search_mode_strata.
+ *
+ * \details
+ *
+ * This configuration wraps the public seqan3::search_cfg::mode. It is used to convert the dynamically configured
+ * search mode into a static mode that can be used within the actual search algorithm implementation.
+ * This is a trick to add the static information to the already existing algorithm configuration containing the
+ * seqan3::search_cfg::mode configuration. Since, the same configuration element cannot be added multiple times a
+ * distinct type is used to forward the actual information. This means that implementors need to use this configuration
+ * instead of seqan3::search_cfg::mode when implementing the algorithm.
+ */
+template <typename search_mode_t>
+//!\cond
+    requires std::same_as<search_mode_t, search_mode_all> ||
+             std::same_as<search_mode_t, search_mode_all_best> ||
+             std::same_as<search_mode_t, search_mode_best> ||
+             std::same_as<search_mode_t, search_cfg::strata>
```

Comment:
This is already the third occurrence of a check for any valid search mode in this file. Maybe it already makes sense to have a local shortcut like
```cpp
template <typename search_mode_t> 
bool is_search_mode = ...
```